manifest_merging/utils.py

```diff
@@ -0,0 +1,215 @@
+        )
+        logging.info(msg)
+        response = s3_client.generate_presigned_url(
+            "get_object",
+            Params={"Bucket": bucket_name, "Key": object_name},
+            ExpiresIn=expiration,
+        )
+    except ClientError as e:
+        print(e)
+        return None
+    # The response contains the presigned URL
+    return response
+
+
+def upload_file_to_s3_and_generate_presigned_url(
+    bucket_name,
+    filename,
+    aws_access_key_id=None,
+    aws_secret_access_key=None,
+    expiration=3600,
+):
+    """
+    Upload and generate presigned url
+
+    Args:
+        bucket_name(str): the bucket name
+        filename(str): the input file needs to be uploaded
+        aws_access_key_id(str): aws access key id
+        aws_secret_access_key(str): aws secret access key
+        expiration(int): expiration time
+
+    Returns:
+        str: presigned url
+    """
+
+    now = datetime.now()
+
+    if filename and bucket_name:
+        current_time = now.strftime("%m_%d_%y_%H:%M:%S")
+        upload_file_key = "{}_{}.log".format(current_time, randomString(6))
+
+        if upload_file(
+            filename,
+            bucket_name,
+            upload_file_key,
+            aws_access_key_id,
+            aws_secret_access_key,
+        ):
+            presigned_url = create_presigned_url(
+                bucket_name, upload_file_key, aws_access_key_id, aws_secret_access_key
+            )
+            return presigned_url
+
+    return None
+
+
+def check_user_permission(access_token, job_requires):
+    """
+    Check if user has permission to run the job or not
+
+    Args:
+        access_token(str): the access token
+        job_requires(dict): requirements so that job can run
+            {
+                "arborist_url": "http://arborist-service",
+                "job_access_req": (
+                    [
+                        {"resource": "/sower", "action": {"service": "job", "method": "access"}},
+                        {"resource": "/programs", "action": {"service": "indexd", "method": "write"}},
+                    ],
+                )
+            }
+    Returns:
+        bool: if user has permission to run the job or not
+        dict: a message log
+    """
+
+    params = {
+        "user": {"token": access_token},
+        "requests": job_requires["job_access_req"],
+    }
+    response = requests.post(
+        "{}/auth/request".format(job_requires["arborist_url"].strip("/")),
+        headers={"content-type": "application/json"},
+        json=params,
+    )
+    if response.status_code != 200:
+        return (
+            False,
+            {"message": "Can not run the job. Detail {}".format(response.json())},
+        )
+
+    elif not response.json()["auth"]:
+        return (False, {"message": "User does not have privilege to run the job"})
+    else:
+        return True, {"message": "OK"}
+
+
+def split_s3_path(s3_path):
```

Comment:
Python already has function for splitting any url. Check it below
https://docs.python.org/3/library/urllib.parse.html
Don't reinvent the wheel

Reply:
While it might not be relevant or even necessary, I think handling fragments and query string parameters with `urllib` could be a little tricky. For example, to correctly handle `s3_path = 's3://awesome-bucket-name/manifest1.tsv?foo=1&bar=2'` or `s3_path = 's3://awesome-bucket-name/manifest1.tsv#hello?foo=1&bar=2'`, code using `urllib` might look like:
```
from urllib.parse import urlparse

def split_s3_path(s3_path):
    parsed = urlparse(s3_path)
    bucket = parsed.netloc
    if parsed.fragment:
        key = parsed.path.lstrip('/') + '#' + parsed.fragment
    elif parsed.query:
        key = parsed.path.lstrip('/') + '?' + parsed.query
    else:
        key = parsed.path.lstrip('/')
    return bucket, key
```
Even if we don't need to handle fragments and query parameters, we still have to `lstrip` the `'/'` that `urllib` leaves.

 As a possible alternative to using `urllib`, what do you think about the following implementation?
```
bucket, key = s3_path.replace("s3://", "").split('/', 1)
```

Reply:
Yeah. That works too!